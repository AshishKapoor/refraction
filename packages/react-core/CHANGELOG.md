# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.9.3](https://github.com/quid/ui-framework/compare/v1.9.2...v1.9.3) (2018-12-11)

**Note:** Version bump only for package @quid/react-core





## [1.9.2](https://github.com/quid/ui-framework/compare/v1.9.1...v1.9.2) (2018-12-11)


### Bug Fixes

* button text should always be black ([5cdfd08](https://github.com/quid/ui-framework/commit/5cdfd08))
* use current background color to style focus ring ([516527e](https://github.com/quid/ui-framework/commit/516527e))





## [1.9.1](https://github.com/quid/ui-framework/compare/v1.9.0...v1.9.1) (2018-12-11)


### Bug Fixes

* properly set JSX pragma at build time ([817eab8](https://github.com/quid/ui-framework/commit/817eab8))





# [1.9.0](https://github.com/quid/ui-framework/compare/v1.8.0...v1.9.0) (2018-12-11)


### Bug Fixes

* button logic ([16d5e71](https://github.com/quid/ui-framework/commit/16d5e71))
* icon default properties ([e3450a9](https://github.com/quid/ui-framework/commit/e3450a9))
* use emotion babel macro ([20c534d](https://github.com/quid/ui-framework/commit/20c534d))
* use Emotion babel macro ([b4e2746](https://github.com/quid/ui-framework/commit/b4e2746))


### Features

* add support for Icon inside Button ([2eccd7f](https://github.com/quid/ui-framework/commit/2eccd7f))





# [1.8.0](https://github.com/quid/ui-framework/compare/v1.7.1...v1.8.0) (2018-12-10)


### Bug Fixes

* move iconic font to [@quid](https://github.com/quid)/theme ([b2a0a97](https://github.com/quid/ui-framework/commit/b2a0a97))


### Features

* new react-core package ([11637bf](https://github.com/quid/ui-framework/commit/11637bf))
