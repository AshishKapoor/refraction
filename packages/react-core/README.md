This package contains all the core UI elements needed to create our
interfaces and the most complex components of this project.

## Installation

```bash
npm install --save @quid/react-core

# or

yarn add @quid/react-core
```

<!-- NPM_ONLY> -->

More documentation is available at https://ui.quid.com

<!-- <NPM_ONLY -->
