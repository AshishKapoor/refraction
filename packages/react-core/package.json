{
  "name": "@quid/react-core",
  "version": "1.14.0",
  "description": "Core React Quid components",
  "main": "dist/index.js",
  "main:umd": "dist/index.umd.js",
  "module": "dist/index.es.js",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/quid/ui-framework.git"
  },
  "scripts": {
    "start": "microbundle watch",
    "prepare": "microbundle build --jsx React.createElement && flow-copy-source --ignore '{__mocks__/*,*.test}.js' src dist",
    "test": "cd ../.. && yarn test --testPathPattern packages/react-core"
  },
  "devDependencies": {
    "flow-copy-source": "^2.0.2",
    "microbundle": "^0.8.3"
  },
  "dependencies": {
    "@emotion/core": "^10.0.4",
    "@emotion/css": "^10.0.4",
    "@emotion/styled": "^10.0.4",
    "@quid/theme": "^1.14.0",
    "color": "^3.1.0",
    "react-router-dom": "^4.3.1"
  },
  "peerDependencies": {
    "react": "15||16"
  }
}
