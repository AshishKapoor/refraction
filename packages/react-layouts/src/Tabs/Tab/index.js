/**
 * Copyright (c) Quid, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// @flow
import * as React from 'react';
import css from '@emotion/css/macro';
import styled from '@emotion/styled/macro';
import { textStyles, withFallback as wf, sizes } from '@quid/theme';

type Props = {
  children: React.Node,
  id: string,
  isActive: boolean,
  onClick: (SyntheticEvent<HTMLElement>) => void,
};

const TabButton = styled.button`
  ${textStyles('normal', 'bold')};
  border: 0;
  font-size: inherit;
  font-family: inherit;
  cursor: pointer;
  background-color: inherit;

  position: relative;
  flex: 1;
  padding: ${sizes.regular};
  line-height: 1;

  &:focus:not(:focus-visible) {
    outline: 0;
  }
  &:focus-visible {
    z-index: 1;
  }

  ${wf(
    props =>
      props.isActive &&
      css`
        background-color: ${props.theme.primaryInverse};
        color: ${props.theme.selected};
      `
  )};
`;

const Tab = styled(({ isActive, id, onClick, children, ...props }: Props) => (
  <li id={id} {...props}>
    <TabButton
      isActive={isActive}
      id={`${id}_button`}
      aria-controls={`${id}_panel`}
      aria-selected={isActive}
      role="tab"
      onClick={onClick}
    >
      {children}
    </TabButton>
  </li>
))`
  flex: 1;
  display: flex;
`;

export default Tab;
