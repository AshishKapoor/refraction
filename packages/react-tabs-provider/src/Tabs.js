// @flow
import * as React from 'react';

export type ContextState = {
  active: ?string,
  select: string => void,
  register: string => void,
};

const noop = () => {};

// prettier-ignore
export const Context = React.createContext/*:: <ContextState> */({
  active: '',
  select: noop,
  register: noop,
});

type Props = {
  children: ({ active: ?string }) => React.Node,
  defaultActive?: string,
  active?: string,
  onSelect?: string => void,
};

type State = {
  registeredTabs: Array<string>,
  active: ?string,
  register: string => void,
  select: string => void,
};

export default class Tabs extends React.Component<Props, State> {
  state = {
    registeredTabs: [],
    active: this.props.defaultActive,
    register: this.register,
    select: this.select,
  };

  register = (name: string) =>
    this.setState(({ registeredTabs }) => ({
      registeredTabs: [...registeredTabs, name],
    }));

  select = (name: string) =>
    this.props.onSelect != null
      ? this.props.onSelect(name)
      : this.setState({
          active: name,
        });

  render() {
    const {
      children,
      active: controlledActive,
      onSelect: controlledOnSelect,
    } = this.props;
    const { active: uncontrolledActive, registeredTabs } = this.state;

    if (
      (controlledActive != null && controlledOnSelect == null) ||
      (controlledOnSelect != null && controlledActive == null)
    ) {
      // eslint-disable-next-line no-console
      console.error(
        'Warning: TabsProvider.Tabs can be used either as a controlled or stateful component, ' +
          'when used as controlled component, make sure to define both `value` and `onSelect` properties to it.'
      );
    }
    const active = controlledActive || uncontrolledActive || registeredTabs[0];

    return (
      <Context.Provider
        value={{ active, register: this.register, select: this.select }}
      >
        {children({ active })}
      </Context.Provider>
    );
  }
}
