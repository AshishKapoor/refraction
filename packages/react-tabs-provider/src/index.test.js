// @flow
import * as React from 'react';
import { render } from 'react-dom';
import { Button } from '@quid/react-core';
import { Tabs, TabList, TabPanel } from '.';
import { Context } from './Tabs';

// TODO: switch to Enzyme once they decide to publish the version with
// createContext support...

it('renders an active TabPanel', () => {
  const wrapper = document.createElement('x-wrapper');
  render(
    <Context.Provider
      value={{ active: 'foo', select: jest.fn(), register: jest.fn() }}
    >
      <TabPanel name="foo">{() => 'foobar'}</TabPanel>
    </Context.Provider>,
    wrapper
  );
  expect(wrapper).toMatchSnapshot();
});

it('renders an inactive TabPanel', () => {
  const wrapper = document.createElement('x-wrapper');
  render(
    <Context.Provider
      value={{ active: 'bar', select: jest.fn(), register: jest.fn() }}
    >
      <TabPanel name="foo">{() => 'foobar'}</TabPanel>
    </Context.Provider>,
    wrapper
  );
  expect(wrapper).toMatchSnapshot();
});

it('renders a TabList with interactable properties', () => {
  const select = jest.fn();
  const register = jest.fn();
  const wrapper = document.createElement('x-wrapper');
  const btn = (
    <button type="button" onClick={() => select('foo')} active="active" />
  );
  render(
    <Context.Provider value={{ active: 'foo', select, register }}>
      <TabList>{({ active, select }) => btn}</TabList>
    </Context.Provider>,
    wrapper
  );
  expect(wrapper).toMatchSnapshot();
  expect(register).not.toBeCalled();

  btn.props.onClick();
  expect(select).toBeCalledWith('foo');
});

it('renders a Tab component that provides the expected context', () => {
  const spySelect = jest.fn((select, name) => select(name));
  let btn = {};
  const wrapper = document.createElement('x-wrapper');
  const tabListFn = jest.fn(({ active, select }) => {
    btn = <button type="button" onClick={() => spySelect(select, 'bar')} />;
    return (
      <React.Fragment>
        <button type="button" onClick={() => select('foo')} active="active" />
        {btn}
      </React.Fragment>
    );
  });

  render(
    <Tabs>
      {() => (
        <div>
          <TabList>{tabListFn}</TabList>
          <TabPanel name="foo">{() => 'foo'}</TabPanel>
          <TabPanel name="bar">{() => 'bar'}</TabPanel>
        </div>
      )}
    </Tabs>,
    wrapper
  );

  btn.props.onClick();

  // click on second button
  expect(spySelect.mock.calls.length).toBe(1);
  expect(spySelect.mock.calls[0][1]).toBe('bar');
});

it('renders an active TabPanel in controlled mode', () => {
  const wrapper = document.createElement('x-wrapper');
  render(
    <Tabs active="foo" onSelect={jest.fn()}>
      {() => (
        <div>
          <TabList>
            {({ select, active }) => (
              <Button
                onClick={() => select && select('a')}
                importance={active === 'a' ? 'primary' : 'secondary'}
              >
                Tab A
              </Button>
            )}
          </TabList>
          <TabPanel name="foo">{() => 'foobar'}</TabPanel>
        </div>
      )}
    </Tabs>,
    wrapper
  );
});

it('renders a Tab component that provides the expected context', () => {
  const spySelect = jest.fn((select, name) => select(name));
  let btn = {};
  const wrapper = document.createElement('x-wrapper');
  const tabListFn = jest.fn(({ active, select }) => {
    btn = <button type="button" onClick={() => spySelect(select, 'bar')} />;
    return (
      <React.Fragment>
        <button type="button" onClick={() => select('foo')} active="active" />
        {btn}
      </React.Fragment>
    );
  });

  const hanleOnSelect = jest.fn();

  render(
    <Tabs active="foo" onSelect={hanleOnSelect}>
      {() => (
        <div>
          <TabList>{tabListFn}</TabList>
          <TabPanel name="foo">{() => 'foo'}</TabPanel>
          <TabPanel name="bar">{() => 'bar'}</TabPanel>
        </div>
      )}
    </Tabs>,
    wrapper
  );

  btn.props.onClick();

  // click on second button
  expect(spySelect.mock.calls.length).toBe(1);
  expect(spySelect.mock.calls[0][1]).toBe('bar');
  expect(hanleOnSelect).toHaveBeenCalled();
});

it('throws an error if not both the controlled mode props are set', () => {
  const wrapper = document.createElement('x-wrapper');
  const handleError = jest.fn();
  global.console = { error: handleError };
  render(
    <Tabs active={undefined} onSelect={jest.fn()}>
      {() => 'foo'}
    </Tabs>,
    wrapper
  );
  expect(handleError).toHaveBeenCalled();
});
