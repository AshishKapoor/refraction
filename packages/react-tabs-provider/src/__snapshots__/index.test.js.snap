// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a TabList with interactable properties 1`] = `
<x-wrapper>
  <button
    active="active"
    type="button"
  />
</x-wrapper>
`;

exports[`renders an active TabPanel 1`] = `
<x-wrapper>
  foobar
</x-wrapper>
`;

exports[`renders an inactive TabPanel 1`] = `<x-wrapper />`;
